File: 백준/Silver/1780. 종이의 개수/종이의 개수.cc
#include<iostream>
using namespace std;

int cnt_1 = 0, cnt0 = 0, cnt1 = 0;
int arr[2200][2200];

bool isSame(int y, int x, int size) { // 모두 같은 수로 이루어져 있는지 확인
	int first = arr[y][x];
	for (int i = y; i < y + size; i++) {
		for (int j = x; j < x + size; j++) {
			if (arr[i][j] != first) {
				return false;
			}
		}
	}
	return true;
}

void div(int y, int x, int size) {
	if (isSame(y, x, size)) { // 모두 같은 수로 이루어져 있는지 확인
		if (arr[y][x] == -1) {
			cnt_1++;
		}
		else if (arr[y][x] == 0) {
			cnt0++;
		}
		else {
			cnt1++;
		}
		return;
	}

	int newSize = size / 3; // 새로운 크기
	for (int i = 0; i < 3; i++) {
		for (int j = 0; j < 3; j++) {
			div(y + i * newSize, x + j * newSize, newSize); // 9등분하여 재귀 호출
		}
	}
}

// 분할정복
int main() {
	int n;
	cin >> n;
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			cin >> arr[i][j];
		}
	}
	div(0, 0, n); // (y, x, size)
	cout << cnt_1 << "\n" << cnt0 << "\n" << cnt1;
	return 0;
}

File: 백준/Silver/1780. 종이의 개수/JAVA/종이의 개수.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    private static int[][] arr;
    private static int[] type; //  -1,0,1
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        arr = new int[N][N];

        for (int i = 0; i < N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < N; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        type = new int[3];
        div(0, 0,N);
        for (int i : type) {
            System.out.println(i);
        }
    }

    private static void div(int y, int x, int size) {
        if (ckeck(y, x, size)) {
            if (arr[y][x] == -1) {
                type[0]++;
            } else if (arr[y][x] == 0) {
                type[1]++;
            } else {
                type[2]++;
            }
            return;
        }

        int newSize = size / 3;
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                div(y+i*newSize,x+j*newSize,newSize);
            }
        }
    }

    private static boolean ckeck(int y, int x, int size) {
        int save = arr[y][x];
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                if (arr[y + i][x + j] != save) {
                    return false;
                }
            }
        }
        return true;
    }
}
